### DISCUSSION TITLE
How to apply a BlockHandler globally to all blocks of a type

### DISCUSSION THREAD

**Discussion by KianBrose** [Minestom/Minestom#1788] [Help]
I recently read through the [World Block](https://wiki.minestom.net/world/blocks) documentation about how to implement block functionality which states that a Handler can be applied to a block object like
```java
Block stone = Block.STONE.withHandler(new DemoHandler());
```
But this would only apply the handler to blocks from the `stone` variable with a setBlock for example, and not blocks that "already exist" within a certain InstanceContainer.
After asking ChatGPT which produced a confidently incorrect answer syntax wise, I saw that **BlockManager** exists and has a `blockManager.registerHandler(NamespaceID,Supplier<BlockHandler>)` which I guess would do this, problem is I have 0 idea of how to use this syntax wise and the IDE isn't really helping either.

I tried
```java
BlockManager blockManager = MinecraftServer.getBlockManager();
blockManager.registerHandler(NamespaceID.from(Key.key("stone.button")), new ButtonHandler());
```
It produces an error and demands me to cast `new ButtonHandler()` to a `Supplier<BlockHandler>` which produces this line:
```java
blockManager.registerHandler(NamespaceID.from(Key.key("stone.button")), (Supplier<BlockHandler>) new ButtonHandler());
```
And this fails to run as well as 
```java
java.lang.ClassCastException: class blockhandlers.ButtonHandler cannot be cast to class 
```
How do I make the BlockHandler `ButtonHandler()` apply to all blocks of a type?

**Comment by reeve567** [comment]
Personally I'd go with an `EventListener` for placing and interaction since you want to get all of them. I haven't been working with Minestom long enough to really understand how they do event handling with nodes, so I've just been doing events like such:

(I use Kotlin but it should be pretty easy to do the same in Java)
``` Kotlin
class BlockPlacing: EventListener<PlayerBlockPlaceEvent> {
	override fun eventType(): Class<PlayerBlockPlaceEvent> {
		...
	}
	
	override fun run(event: PlayerBlockPlaceEvent): EventListener.Result {
		...
	}
}


globalEventHandler.addListener(BlockPlacing())
```

**Reply by KianBrose** [reply]
Wouldn't this just affect new placed blocks? What I'm looking for is a way to make  a handler already apply to pre placed blocks without having to place them again

**Reply by reeve567** [reply]
Yes, this would listen for new placed blocks, then you can do another with `PlayerBlockInteractEvent`, that will listen for any blocks already placed to interact with

**Reply by reeve567** [reply]
Could even just only use the `PlayerBlockInteractEvent` to listen for placing as well, but I figured that'd make it a little more complicated

**Reply by reeve567** [reply]
Also reading back on this, what does your ButtonHandler extend right now?
If casting is your issue, then you should just need to implement `Supplier<BlockHandler>`

**Reply by KianBrose** [reply]
It does work as you say, but my particular use case is that a lot of blocks within the map already exist, and I desire to give them functionality without having to re-place them again hence the issue

**Reply by KianBrose** [reply]
Now that I re-read, I will have a look at the PlayerBlockInteractEvent and give it a shot, in theory it should suffice
